protein_helper function

yuyuan871111 opened the pull request:
First version of the protein helper function (not yet finished)
TODO:

- [ ] the decision of input format of molecule
- [ ] fix_molecule_bond_orders function

cbouy commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,1958 @@
+# HID (neutral, hydrogen at ND1, might couple to HEME at NE2),
+# ILE, LEU, LYS, MET, PRO, PHE, SER, TYR, THR, TRP, VAL
+# HOH (water)
+_STANDARD_AA = """\
```
I'd just put this in a separate `data/` file instead

cbouy commented on prolif/data/implicitHbond/1s2g__1__1.A_2.C__1.D/chain_mapping.json:
I know this is a WIP but be careful with the extra files that you commit!
You can specify files to ignore in `.git/info/exclude` (equivalent of `.gitignore` but specific to your local machine)

talagayev commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,1958 @@
+        # fix molecule bond orders with templates
+
+
+# The below code contains functions for fixing molecule bond orders.
```
Agree, SMILES should be easier from a users perspective. From the `input_topology` I am also thinking if it is not better for the users to have just the `ProLIF Molecule` since then it is easier with the `convert_to_standard_resname`, since I can imagine that it can happen that users would try to directly use their `Amber` or `CHARMM` topology as an input directly quite often, which will lead to them getting the `TypeError`. Maybe use MDAnalysis in the backend for the conversion of the input topology to a Prolif Molecule @cbouy?

yuyuan871111 replied:
Now, users can use the SMILES as a template.

talagayev commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,520 @@
+from prolif.molecule import Molecule
+from prolif.residue import Residue
+
+all_resnames_dict = {
```
Here I would possibly tend to move the `all_resnames_dict` in a separate file similar to `constants.py` and use it from there

cbouy commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,560 @@
+            [
+                atom
+                for atom in residue.GetAtoms()
+                if atom.GetSymbol().strip() != "H"
```
no need for `strip` on `GetSymbol`, but I'd rather see `GetAtomicNum` instead
```suggestion
                if atom.GetAtomicNum() != 1
```

cbouy commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,560 @@
+        int
+            The number of heavy atoms in the residue.
+        """
+        terminal_oxygen = {
```
this should be a constant

cbouysset replied:
Sorry should have been more specific, there's no way in Python to enforce a var to be constant (immutable), so the conventional way is to use an uppercase var name e.g. `TERMINAL_OXYGEN`.

Also there's no need for it to be re-declared inside the function at every call of `n_residue_heavy_atoms` so you can pull it out of here and put it with the other constants

cbouy commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,560 @@
+# The below code contains functions for fixing molecule bond orders.
+# The idea is based on pdbinf: https://github.com/OpenFreeEnergy/pdbinf/tree/main
+# A user can provide a custom CIF file with the standard amino acid.
+def _block_decompose(data_block: list) -> tuple:
```
this function and `cif_parser_lite` should be in a separate `cif` module. Maybe we should have a `io` folder to contain this and the protein helper module

cbouy commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,560 @@
+
+
+def strip_bonds(m: Chem.Mol) -> Chem.Mol:
+    em = AllChem.EditableMol(m)
+
+    for b in m.GetBonds():
+        em.RemoveBond(b.GetBeginAtomIdx(), b.GetEndAtomIdx())
```
Use `Chem.RWMol` instead
```suggestion
    with Chem.RWMol(m) as em:
        for b in m.GetBonds():
            em.RemoveBond(b.GetBeginAtomIdx(), b.GetEndAtomIdx())
```

cbouy commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,560 @@
+
+    logger.debug(f"assigning intra props for {reference_block['name']}")
+
+    em = AllChem.EditableMol(mol)
```
same as above with `RWMol`

cbouy commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,560 @@
+# HID (neutral, hydrogen at ND1, might couple to HEME at NE2),
+# ILE, LEU, LYS, MET, PRO, PHE, SER, TYR, THR, TRP, VAL
+# HOH (water)
+STANDARD_AA = cif_parser_lite(_STANDARD_AA)
```
this should be in the constants file, or at the top of this file if there's a circular import

cbouy commented on prolif/protein_helper.py:
```diff
@@ -0,0 +1,560 @@
+                    stacklevel=2,
+                )
+            # fix the bond orders
+            new_residues.append(self.fix_molecule_bond_orders(residue, **kwargs))
```
one thing to look at in the future is whether we need to run this for every single residue or if we can detect early which ones need to be fixed to save some time

cbouysset commented on prolif/constants.py:
```diff
@@ -0,0 +1,285 @@
+_STANDARD_AA = Path(str(datapath / "standard_aa.cif")).read_text()
+STANDARD_AA = cif_parser_lite(_STANDARD_AA)
+# --------- the above is originally from openmm and pdbinf ---------
+breakpoint()
```
don't forget to clean this up!

cbouysset commented on prolif/io/protein_helper.py:
```diff
@@ -105,13 +100,13 @@ def forcefield_guesser(
+        if AMBER_POOL.intersection(conv_resnames):
             return "amber"
-        if len(charmm_pool.intersection(conv_resnames)) != 0:
+        if len(CHARMM_POOL.intersection(conv_resnames)) != 0:
```
you can remove the `len` and `!=0` here and below as well

cbouysset commented on prolif/io/protein_helper.py:
```diff
@@ -207,7 +192,7 @@ def n_residue_heavy_atoms(residue: Residue) -> int:
         Returns
         -------
-        int
+        const int
```
you can remove the `const`, sorry about the confusion!

cbouysset reviewed (commented):
Looks much better, thanks!

cbouy commented on prolif/io/protein_helper.py:
```diff
@@ -0,0 +1,594 @@
+
+    """
+
+    def __init__(self, input_topology: Molecule | str | Path):
```
Thinking about this a bit more, I think we should invert the input params between the initialization and the standardize method. The protein-helper should be configured with the templates, and then ran on a specific input, rather than the other way around

cbouy commented on prolif/io/protein_helper.py:
```diff
@@ -0,0 +1,594 @@
+            # if templates is a list, check if it contains dicts
+            if not all(isinstance(t, dict) for t in templates):
+                raise TypeError("Templates must be a dict, a list of dicts or None.")
+            templates.extend([STANDARD_AA])
```
Best to avoid directly modifying the list provided by the user:
```suggestion
            templates = [*templates, STANDARD_AA]
```

cbouy commented on tests/test_io.py:
```diff
@@ -0,0 +1,426 @@
+
+def test_cif_template_reader() -> None:
+    """Test the CIF template reader."""
+    from prolif.io.cif import cif_template_reader
```
same here

cbouy commented on tests/test_io.py:
```diff
@@ -0,0 +1,426 @@
+
+def test_xml_parse_altnames(XML_TEST_DATA: str) -> None:
+    """Test the XML parsing of alternative names for residues and atoms."""
+    from prolif.io.xml import parse_altnames
```
same here

cbouy commented on tests/test_io.py:
```diff
@@ -0,0 +1,426 @@
+
+        # check two residue are equal and bond orders are fixed
+        for at1, at2 in zip(
+            fixed_mol.GetAtoms(), fixed_mol_custom.GetAtoms(), strict=False
```
strict should be set to True, we don't expect the number of atoms to be different here
```suggestion
            fixed_mol.GetAtoms(), fixed_mol_custom.GetAtoms(), strict=True
```

cbouy commented on tests/test_io.py:
```diff
@@ -0,0 +1,426 @@
+        ):
+            assert at1.GetSymbol() == at2.GetSymbol()
+
+        bond1_info, bond2_info = [], []
```
not a fan of this test, I'd just check that the number of bonds is the same, and then just iterate over the bonds of either, get the corresponding bond in the other mol, and check that the type is the same, something like:
```python
for bond1 in fixed_mol.GetBonds():
    bond2 = fixed_mol_custom.GetBondBetweenAtoms(bond1.GetBeginAtomIdx(), bond1.GetEndAtomIdx())
   assert bond1.GetBondType() is bond2.GetBondType()
```
in the end it does the same but is much simpler

cbouy commented on tests/test_io.py:
```diff
@@ -0,0 +1,426 @@
+        from prolif.io.protein_helper import ProteinHelper
+
+        # Test with a known forcefield
+        forcefield = ProteinHelper.forcefield_guesser({"HSD"})
```
use `pytest.mark.parametrize` when you run the same logic on different inputs instead of repeating code. One var for the residue name, the other for the forcefield

cbouy commented on tests/test_io.py:
```diff
@@ -0,0 +1,426 @@
+        forcefield = ProteinHelper.forcefield_guesser({"CYS"})
+        assert forcefield == "unknown"
+
+    def test_standardize_protein(
```
this test does too many different things. Tests should be written as functions and have a single responsibility e.g. one test for checking if a warning is emitted, one test for checking if the output is as expected...etc.

cbouy commented on tests/test_io.py:
```diff
@@ -0,0 +1,426 @@
+            # test the function
+            em_fixed = _assign_intra_props_lone_H(em)
+
+        for bond1, bond2 in zip(mol.GetBonds(), em_fixed.GetBonds(), strict=False):
```
```suggestion
        for bond1, bond2 in zip(mol.GetBonds(), em_fixed.GetBonds(), strict=True):
```

cbouy reviewed (changes requested):
Nice to see the support for SMILES templates! I realised that the way the ProteinHelper is called is a bit weird, it should be configured with the templates, and then you can pass your input topology to the relevant methods.
For tests you should avoid importing within the test, and try to use fixtures more to avoid duplicating function calls with different values

talagayev commented on tests/test_io.py:
```diff
@@ -0,0 +1,426 @@
+
+
+def test_cif_parser_lite(CIF: str, STANDARD_AA: str) -> None:
+    from prolif.io.cif import cif_parser_lite
```
agree, it also then reduces the duplicates, since some `cif_template_reader` it is imported twice, since you import it in two different tests, so if you move it to the top you reduce the amount of duplicates

yuyuan871111 commented:
Hi @cbouy @cbouysset @talagayev ,

I have updated `test_constants.py` and removed some redundant data (also updated the readme to let user know where the data comes from).

Please let me know if any issues need to be addressed.
